File: examples/python/list_plans.py
#!/usr/bin/env python3
"""List all plans with status.

Usage: ./list_plans.py [repo-name]
"""
import sys
import re
from pathlib import Path


def parse_frontmatter(content: str) -> dict:
    """Extract YAML frontmatter from markdown."""
    match = re.match(r"^---\n(.*?)\n---", content, re.DOTALL)
    if not match:
        return {}

    data = {}
    for line in match.group(1).split("\n"):
        if ":" in line:
            key, value = line.split(":", 1)
            data[key.strip()] = value.strip()
    return data


def list_plans(repo: str = None) -> list:
    """Find and parse all plan files."""
    plans_dir = Path("plans")
    if not plans_dir.exists():
        return []

    pattern = f"{repo}/*.md" if repo else "*/*.md"
    plans = []

    for path in plans_dir.glob(pattern):
        content = path.read_text()
        meta = parse_frontmatter(content)
        if meta.get("id"):
            plans.append({
                "id": meta.get("id"),
                "status": meta.get("status", "unknown"),
                "repo": meta.get("repo", path.parent.name),
                "file": str(path),
            })

    return sorted(plans, key=lambda p: p["status"])


if __name__ == "__main__":
    repo = sys.argv[1] if len(sys.argv) > 1 else None
    plans = list_plans(repo)

    if not plans:
        print("No plans found.")
        sys.exit(0)

    # Simple table output
    print(f"{'ID':<10} {'Status':<12} {'Repo':<20} File")
    print("-" * 60)
    for p in plans:
        print(f"{p['id']:<10} {p['status']:<12} {p['repo']:<20} {p['file']}")

File: examples/typescript/list-plans.ts
#!/usr/bin/env npx tsx
/**
 * List all plans with status.
 * Usage: npx tsx list-plans.ts [repo-name]
 */
import { readdirSync, readFileSync, existsSync } from "fs";
import { join } from "path";

interface Plan {
  id: string;
  status: string;
  repo: string;
  file: string;
}

function parseFrontmatter(content: string): Record<string, string> {
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) return {};

  const data: Record<string, string> = {};
  for (const line of match[1].split("\n")) {
    const idx = line.indexOf(":");
    if (idx > 0) {
      data[line.slice(0, idx).trim()] = line.slice(idx + 1).trim();
    }
  }
  return data;
}

function listPlans(repo?: string): Plan[] {
  const plansDir = "plans";
  if (!existsSync(plansDir)) return [];

  const plans: Plan[] = [];
  const repos = repo ? [repo] : readdirSync(plansDir);

  for (const r of repos) {
    const repoDir = join(plansDir, r);
    if (!existsSync(repoDir)) continue;

    for (const file of readdirSync(repoDir)) {
      if (!file.endsWith(".md")) continue;

      const path = join(repoDir, file);
      const content = readFileSync(path, "utf-8");
      const meta = parseFrontmatter(content);

      if (meta.id) {
        plans.push({
          id: meta.id,
          status: meta.status || "unknown",
          repo: meta.repo || r,
          file: path,
        });
      }
    }
  }

  return plans.sort((a, b) => a.status.localeCompare(b.status));
}

// CLI
const [, , repo] = process.argv;
const plans = listPlans(repo);

if (plans.length === 0) {
  console.log("No plans found.");
  process.exit(0);
}

// Table output
console.log(
  `${"ID".padEnd(10)} ${"Status".padEnd(12)} ${"Repo".padEnd(20)} File`
);
console.log("-".repeat(60));
for (const p of plans) {
  console.log(
    `${p.id.padEnd(10)} ${p.status.padEnd(12)} ${p.repo.padEnd(20)} ${p.file}`
  );
}
